### PR TITLE
Add cma/fmin2

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by teytaud** [comment]
Maybe deprecated by https://github.com/facebookresearch/nevergrad/pull/1254 which directly adds a bet-and-run in NGOpt.

File: nevergrad/optimization/recastlib.py [hunk 1]
```diff
@@ -28,11 +29,23 @@ def __init__(
...
+            "SLSQP",
+            "Powell",
+        ], f"Unknown method '{method}'"
         self.method = method
         self.random_restart = random_restart
+        # The following line rescales to [0, 1] if fully bounded.
+        if method == "CmaFmin2" and p.helpers.Normalizer(self.parametrization).fully_bounded:
+            self._normalizer = p.helpers.Normalizer(self.parametrization)
+        else:
+            self._normalizer = None
```
Comment (jrapin) [inline nevergrad/optimization/recastlib.py:48]: Initialization is heavy so better compute it only once
```suggestion
        self._normalizer: tp.Optional[p.helpers.Normalizer] = None
        if method == "CmaFmin2":
             normalizer = p.helpers.Normalizer(self.parametrization)
             if normalizer.fully_bounded:
            self._normalizer = normalizer
```

File: nevergrad/optimization/recastlib.py [hunk 2]
```diff
@@ -63,21 +76,54 @@ def _optimization_function(self, objective_function: tp.Callable[[tp.ArrayLike],
...
-                best_x if not self.random_restart else self._rng.normal(0.0, 1.0, self.dimension),
-                method=self.method,
-                options=options,
-                tol=0,
-            )
-            if res.fun < best_res:
-                best_res = res.fun
-                best_x = res.x
+            if self.method == "CmaFmin2":
+                if p.helpers.Normalizer(self.parametrization).fully_bounded:
```
Comment (jrapin) [inline nevergrad/optimization/recastlib.py:80]:
```suggestion
                if self._normalizer is not None:
```

File: nevergrad/optimization/recastlib.py [hunk 3]
```diff
@@ -63,21 +76,54 @@ def _optimization_function(self, objective_function: tp.Callable[[tp.ArrayLike],
...
+                    options["bounds"] = [0.0, 1.0]
+
+                def cma_objective_function(data):
+                    # Hopefully the line below does nothing if unbounded and rescales from [0, 1] if bounded.
+                    if self._normalizer is not None:
+                        data = self._normalizer.backward(np.asarray(data, dtype=np.float32))
+                    return objective_function(data)
+
+                # cma.fmin2(objective_function, [0.0] * self.dimension, [1.0] * self.dimension, remaining)
+                x0 = [0.5] * self.dimension
```
Comment (jrapin) [inline nevergrad/optimization/recastlib.py:91]: only if normalized?

```suggestion
                x0 = 0.5 * np.ones(self.dimension)
```

File: nevergrad/optimization/recastlib.py [hunk 4]
```diff
@@ -63,21 +76,54 @@ def _optimization_function(self, objective_function: tp.Callable[[tp.ArrayLike],
...
+                    res = cma.fmin(
+                        cma_objective_function,
+                        x0=x0,
+                        sigma0=0.2,
+                        options=options,
+                        restarts=9,
+                    )
+                    x0 = list(
+                        0.5 + np.random.uniform() * (2.0 * np.random.uniform(size=self.dimension) - 1) / 2.0
+                    )
```
Comment (jrapin) [inline nevergrad/optimization/recastlib.py:104]: i dont get this

Comment (teytaud) [reply nevergrad/optimization/recastlib.py:104]: Just randomly drawing a restart in the domain [0, 1]^d when bounded ?

File: nevergrad/optimization/recastlib.py [hunk 5]
```diff
@@ -57,21 +71,54 @@ def _optimization_function(
...
+                    if self._normalizer is not None:
+                        data = self._normalizer.backward(np.asarray(data, dtype=np.float32))
+                    return objective_function(data)
+
+                # cma.fmin2(objective_function, [0.0] * self.dimension, [1.0] * self.dimension, remaining)
+                x0 = 0.5 * np.ones(self.dimension)
+                num_calls = 0
+                while budget - num_calls > 0:
+                    options = {"maxfevals": budget - num_calls, "verbose": -9}
+                    if p.helpers.Normalizer(self.parametrization).fully_bounded:
```
Comment (jrapin) [inline nevergrad/optimization/recastlib.py:87]: Instantiating a normalizer is slow, do it only in init
```suggestion
                    if self._normalizer is not None:
```

File: nevergrad/optimization/recastlib.py [hunk 6]
```diff
@@ -57,21 +71,52 @@ def _optimization_function(
...
+                        # Tell CMA to work in [0, 1].
+                        options["bounds"] = [0.0, 1.0]
+                    res = cma.fmin(
+                        cma_objective_function,
+                        x0=x0,
+                        sigma0=0.2,
+                        options=options,
+                        restarts=9,
+                    )
+                    x0 = 0.5 + np.random.uniform() * (2.0 * np.random.uniform(size=weakself.dimension) - 1) / 2.0
```
Comment (jrapin) [inline nevergrad/optimization/recastlib.py:97]:
```suggestion
                    x0 = 0.5 + np.random.uniform() * np.random.uniform(low =-0.5, high=0.5, size=self.dimension)
```

File: nevergrad/optimization/recastlib.py [hunk 7]
```diff
@@ -57,21 +71,52 @@ def _optimization_function(
...
+                        # Tell CMA to work in [0, 1].
+                        options["bounds"] = [0.0, 1.0]
+                    res = cma.fmin(
+                        cma_objective_function,
+                        x0=x0,
+                        sigma0=0.2,
+                        options=options,
+                        restarts=9,
+                    )
+                    x0 = 0.5 + np.random.uniform() * np.random.uniform(low =-0.5, high=0.5, size=self.dimension)
```
Comment (jrapin) [inline nevergrad/optimization/recastlib.py:97]:
```suggestion
                    x0 = 0.5 + np.random.uniform() * np.random.uniform(low=-0.5, high=0.5, size=self.dimension)
```

File: nevergrad/optimization/recastlib.py [hunk 8]
```diff
@@ -57,21 +71,57 @@ def _optimization_function(
...
-            )
-            if res.fun < best_res:
-                best_res = res.fun
-                best_x = res.x
+            # options: tp.Dict[str, tp.Any] = {} if self.budget is None else {"maxiter": remaining}
+            if weakself.method == "CmaFmin2":
+
+                def cma_objective_function(data):
+                    # Hopefully the line below does nothing if unbounded and rescales from [0, 1] if bounded.
+                    if self._normalizer is not None:
```
Comment (jrapin) [inline nevergrad/optimization/recastlib.py:79]:
```suggestion
                    if weakself._normalizer is not None:
```

File: nevergrad/optimization/recastlib.py [hunk 9]
```diff
@@ -57,21 +71,57 @@ def _optimization_function(
...
+                        data = self._normalizer.backward(np.asarray(data, dtype=np.float32))
+                    return objective_function(data)
+
+                # cma.fmin2(objective_function, [0.0] * self.dimension, [1.0] * self.dimension, remaining)
+                x0 = 0.5 * np.ones(weakself.dimension)
+                num_calls = 0
+                while budget - num_calls > 0:
+                    options = {"maxfevals": budget - num_calls, "verbose": -9}
+                    if self._normalizer is not None:
```
Comment (jrapin) [inline nevergrad/optimization/recastlib.py:88]:
```suggestion
                    if weakself._normalizer is not None:
```

**Pull request merged**